### PR TITLE
Change analysis.py to numpydoc style

### PR DESCRIPTION
**Pull request by KybernetikJo**
This PR refactors the `Slycot/slycot/analysis.py` in order to be closer to the numpydoc style as discussed in #100.
Should be part of #204.

(finished 0 of 18 routines)

| Routine name| Finished |Status  description  |
|---------------------|--------------|-----------------------------|
| ab01nd  | <ul><li>[ ] </li></ul> | *3 |
| ab05md  | <ul><li>[ ] </li></ul> | *3   |
| ab05nd  | <ul><li>[ ] </li></ul> |  *3 |
| ab07nd  | <ul><li>[ ] </li></ul> |  *3 |
| ab08nd  | <ul><li>[ ] </li></ul> |  *3 |
| ab08nz  | <ul><li>[ ] </li></ul> |  *3 |
| ab09ad  | <ul><li>[ ] </li></ul> |  *3 |
| ab09ax  | <ul><li>[ ] </li></ul> |  *3 |
| ab09bd  | <ul><li>[ ] </li></ul> |  *3 |
| ab09md  | <ul><li>[ ] </li></ul> |  *3 |
| ab09nd  | <ul><li>[ ] </li></ul> |  *3 |
| ab13bd  | <ul><li>[ ] </li></ul> |  *3 |
| ab13dd  | <ul><li>[ ] </li></ul> |  *3 |
| ab13ed  | <ul><li>[ ] </li></ul> |  *3 |
| ab13fd  | <ul><li>[ ] </li></ul> |  *3 |
| ab13md  | <ul><li>[ ] </li></ul> |  *3 |
| ag08bd  | <ul><li>[ ] </li></ul> |  *3 |

Status description:
1. numpydoc docstring sections, fixed
2. parameters data types, fixed and 1 fixed
3. optional and default values, fixed and 2 fixed
4. ...., fixed and 3 fixed

At least every slycot routine should have the following [numpydoc docstring sections](https://numpydoc.readthedocs.io/en/latest/format.html#sections)
- Parameters
- Returns
- Raises or Warns

Scipy often uses the numpydoc docstring sections 
- Parameters
- Returns
- See Also
- Notes
- References
- Examples
- (see [scipy.signal.place_poles](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.place_poles.html#scipy.signal.place_poles) or [scipy.linalg.solve_continuous_are](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve_continuous_are.html#scipy.linalg.solve_continuous_are))

Some parts may be related to the final decision on #199, #200.
- Behavior of `intent(in,out)` vs `intent(in,out,copy)` must be investigated again. (see discussion #200)
- Parameters: ndarray or array_like, depending on `intent(in,out)` vs `intent(in,out,copy)`?
- Returns: ndarray or array_like, depending on `intent(in,out)` vs `intent(in,out,copy)`?
-  [numpydoc, array_like](https://numpydoc.readthedocs.io/en/latest/format.html) "array_like : For functions that take arguments which can have not only a type ndarray, but also types that can be converted to an ndarray (i.e. scalar types, sequence types), those arguments can be documented with type array_like."

Further, discussing is needed.

### CODE REVIEW

File: slycot/analysis.py [hunk 1]
```diff
@@ -139,76 +144,84 @@ def ab01nd(n, m, A, B, jobz='N', tol=0, ldwork=None):
...
+        The number of input variables for the first system. m1 > 0.
+    p1 : int
+        The number of output variables from the first system and the number
+        of input variables for the second system. p1 > 0.
+    n2 : int
+        The number of state variables in the second system, i.e. the order
+        of the matrix A2.  n2 > 0.
+    p2 : int
+        The number of output variables from the second system. p2 > 0.
+    A1 : (n1,n1) array_like
```
Comment (bnavigator) [inline slycot/analysis.py:168]: nitpick: there is supposed to be a space after the comma

**Comment by KybernetikJo** [comment]
I have done the main refactoring, which includes mainly:
- numpydoc sections `(------)`
- `name : data types` `(A : array_like)`
- optional and default values, 

I think some default values are still missing, though.

-----

I have added the lines in every routine

Raises
------
SlycotParameterError
     :info = -i: the i-th argument had an illegal value;

but I am not completely sure about that.

I have explored the slycot exceptions mechanism in
[https://github.com/KybernetikJo/slycot-examples/tree/main/unstable/slycot_exceptions](https://github.com/KybernetikJo/slycot-examples/tree/main/unstable/slycot_exceptions).

I am not completely sure that the SlycotResultWarning works correctly.

I guess the best way to make sure that the docsting parts Raises & Warns work would be to write pytests.

Should we refactor the pytests in this PR?
I mean all the analysis pytests. If so, I would suggest that each wrapper has its own test:
- test_ab01nd.py
- test_ab05md.py
- ...

**Comment by KybernetikJo** [comment]
A rebase and squash of commits would also be good.

**Comment by bnavigator** [comment]
The `info` parameter depends on the Fortran routine, not all have it and not all define it the same way. Check the SLICOT docs.

> The best way to make sure that the docsting parts Raises & Warns work would be to write pytests.

We have master/slycot/tests/test_analysis.py for that.

Which is separate from making sure that the wrapper function correctly raises the advertised exceptions. Those tests would need to be written, which would also increase code coverage.

File: slycot/analysis.py [hunk 2]
```diff
@@ -1831,75 +1901,85 @@ def ag08bd(l,n,m,p,A,E,B,C,D,equil='N',tol=0.0,ldwork=None):
...
+    infz : (n+1, ) ndarray
+        The leading DINFZ elements of infz contain information
+        on the infinite elementary divisors as follows:
+        the system has infz(i) infinite elementary divisors of
+        degree i in the Smith form, where i = 1,2,...,DINFZ.
+    kronr : (n+m+1, ) ndarray
+        The leading NKROR elements of this array contain the
+        right Kronecker (column) indices.
+    infe : (1+min(l+p,n+m), ) ndarray
+        The leading NINFE elements of infe contain the
```
Comment (bnavigator) [inline slycot/analysis.py:1976]: NINFE is hidden in the output

Comment (bnavigator) [reply slycot/analysis.py:1976]: SO are NKROR and DINFZ

File: slycot/analysis.py (hunk 1 above)
Comment (KybernetikJo) [reply slycot/analysis.py]: That one has already been fixed.

File: slycot/analysis.py (hunk 2 above)
Comment (KybernetikJo) [reply slycot/analysis.py]: DINF, NKROR, NINFE, NKROL in the docstring have been deleted, ag08bd.

File: slycot/analysis.py [hunk 3]
```diff
@@ -221,69 +220,73 @@ def ab05md(n1,m1,p1,n2,p2,A1,B1,C1,D1,A2,B2,C2,D2,uplo='U'):
...
+        matrix C for the cascaded system.
+    D : (p2, m1) ndarray
+        The leading p2-by-m1 part of this array contains the input/output
+        matrix D for the cascaded system.
+
+    Notes
+    -----
         The implemented methods rely on accuracy enhancing square-root or
         balancing-free square-root techniques.
         The algorithms require less than 30N^3  floating point operations.
```
Comment (bnavigator) [inline slycot/analysis.py:292]: Unindent here?

File: slycot/analysis.py [hunk 4]
```diff
@@ -657,77 +676,79 @@ def ab09ad(dico,job,equil,n,m,p,A,B,C,nr=None,tol=0,ldwork=None):
...
+        precision (see LAPACK Library Routine DLAMCH) and
+        ``HNORM(A,B,C)`` is the Hankel norm of the system (computed
+        in ``HSV(1)``);
+        If on input ``nr is None``, `nr` is equal to the number of Hankel
+        singular values greater than ``MAX(tol,n*eps*HNORM(A,B,C))``.
+    Ar : (nr, nr) ndarray
+        This array contains the state dynamics matrix `Ar` of the reduced
+        order system.
+    Br : (nr, m) ndarray
+        Tthis array contains the input/state matrix `Br` of the reduced
```
Comment (bnavigator) [inline slycot/analysis.py:743]:
```suggestion
        This array contains the input/state matrix `Br` of the reduced
```

File: slycot/analysis.py [hunk 5]
```diff
@@ -783,81 +804,83 @@ def ab09ax(dico,job,n,m,p,A,B,C,nr=None,tol=0.0,ldwork=None):
...
+        matrix B of the system.
+    C : (p, n) array_like
+        The leading p-by-n part of this array must contain the
+        state/output matrix C of the system.
+    nr : int, optional
+        `nr` is the desired order of the resulting reduced order
+        system.  ``0 <= nr <= n``. Automatically determined by `tol` if
+        ``nr is None`` and returned. See return object `nr`.
+        Default is None.
+    tol : double precision, optional
```
Comment (bnavigator) [inline slycot/analysis.py:833]:
```suggestion
    tol : float, optional
```

File: slycot/analysis.py [hunk 6]
```diff
@@ -783,81 +804,83 @@ def ab09ax(dico,job,n,m,p,A,B,C,nr=None,tol=0.0,ldwork=None):
...
+        If ``nr is None``, `tol`contains the tolerance for determining the
```
Comment (bnavigator) [inline slycot/analysis.py:834]:
```suggestion
        If `nr` is `None`, `tol` contains the tolerance for determining the
```

File: slycot/tests/test_ab05md.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+from scipy import linalg, signal
+
+from slycot import analysis
+
+
+class Test_ab05md:
+
+    @pytest.mark.skip("not implemented yet, TODO")
+    def test_ab05md():
+        pass
```
Comment (bnavigator) [inline slycot/tests/test_ab05md.py:13]: I'm not a fan of these. They pollute the report about skipped tests but increase collection time for no real benefit

File: slycot/tests/test_analysis.py [hunk 8]
```diff
@@ -1,44 +0,0 @@
...
-                                                               'Nr': 2,
-                                                               'alpha': -0.1}),
-     (analysis.ab13bd, SlycotArithmeticError, 6,              {'dico': 'C'}),
-     (analysis.ab13bd, SlycotResultWarning,   ((1, 0),),      {}),
-     (analysis.ab13dd, SlycotArithmeticError, 4,              {}),
-     (analysis.ab13ed, SlycotArithmeticError, 1,              {}),
-     (analysis.ab13fd, SlycotArithmeticError, (2,),           {}),
-     (analysis.ab13fd, SlycotResultWarning,   (1,),           {})))
-def test_ab_docparse(fun, exception_class, erange, checkvars):
-    assert_docstring_parse(fun.__doc__,  exception_class, erange, checkvars)
```
Comment (bnavigator) [inline slycot/tests/test_analysis.py:42]: Please keep the parametrization together. Splitting them out is unnecessary code repetition.

**Comment by bnavigator** [comment]
This is huge work. Thanks for still working on these. I'd like to do one or two more iterations before we merge, though.

**Comment by KybernetikJo** [comment]
@bnavigator Thanks for your comments. I will look into your feedback.

As I have said in https://github.com/python-control/Slycot/pull/211#issuecomment-1699685765, I am not sure about splitting the tests into single files. However, to have all tests in one file is also suboptimal.

We could organize the tests after the first 4 characters. The docstring-info tests (former test_analysis.py), however, in an extra file. 
The files would then be organized as follows:

test_analysis_info.py (former test_analysis.py)
test_ab01.py
test_ab04.py
test_ab05.py
...
test_ab13.py
test_ag08.py

Or are there other options?
Any suggestions are welcome.

File: slycot/tests/test_ab05md.py (hunk 7 above)
Comment (KybernetikJo) [reply slycot/tests/test_ab05md.py:13]: So this should be deleted again?
Then we have a wrapper, but no test.

Is that OK?

File: slycot/analysis.py (hunk 3 above)
Comment (KybernetikJo) [reply slycot/analysis.py:292]: Yup, was so before https://github.com/python-control/Slycot/blob/master/slycot/analysis.py .

I will change it to be unindent.

**Comment by KybernetikJo** [comment]
Another design decision concerns my use of test classes, even if there is only one test.

```python
class Test_ab01nd:

    def test_ab01nd(self):
```

One reason for the use of classes was the splitting of the test_analysis.py file. https://github.com/python-control/Slycot/pull/206#discussion_r1443859406

Should we keep this?

File: slycot/tests/test_ab05md.py (hunk 7 above)
Comment (bnavigator) [reply slycot/tests/test_ab05md.py:13]: I'd rather have no test than a test that does nothing. We already have the missing coverage as indication that there should be one implemented

**Comment by bnavigator** [comment]
Yes, please keep the docstring tests in one single file.

Please remove the classes. They're just duplicating the name for no additional functionality.

File: slycot/analysis.py (hunk 6 above)
Comment (KybernetikJo) [reply slycot/analysis.py:834]: It is true, there is a difference between ``monospace`` (not for variables) and `single back-ticks` (for variable, module, function) in the numpydoc style (https://numpydoc.readthedocs.io/en/latest/format.html#common-rest-concepts).

```markdown
For paragraphs, indentation is significant and indicates indentation in the output. New paragraphs are marked with a blank line.

Use *italics*, **bold** and ``monospace`` if needed in any explanations (but not for variable names and doctest code or multi-line code). Variable, module, function, and class names should be written between single back-ticks (`numpy`).

A more extensive example of reST markup can be found in [this example document](http://docutils.sourceforge.net/docs/user/rst/demo.txt); the [quick reference](http://docutils.sourceforge.net/docs/user/rst/quickref.html) is useful while editing.

Line spacing and indentation are significant and should be carefully followed.
```

I think this kind of numpydoc correction should be done after the merge of the four basic clean-ups (math, transform, analysis, synthesis).  The use of `` `` and ` ` is all over slycot.

I would prefer to fix the pytest file-structure first and the main numpydoc organization. After the merge of all 4 open PR's we can do smaller numpydoc fixes.

File: slycot/tests/test_ab05md.py (hunk 7 above)
Comment (KybernetikJo) [reply slycot/tests/test_ab05md.py:13]: OK

**Comment by KybernetikJo** [comment]
What do you think about the file structure? https://github.com/python-control/Slycot/pull/206#issuecomment-1879847670

File: slycot/analysis.py (hunk 4 above)
Comment (KybernetikJo) [reply slycot/analysis.py]: Done.

File: slycot/analysis.py (hunk 3 above)
Comment (KybernetikJo) [reply slycot/analysis.py]: Done.

File: slycot/analysis.py (hunk 5 above)
Comment (KybernetikJo) [reply slycot/analysis.py]: Done.

File: slycot/analysis.py (hunk 6 above)
Comment (bnavigator) [reply slycot/analysis.py:834]: Okay, but please add the space after `tol` now.

**Comment by bnavigator** [comment]
> What do you think about the file structure? [#206 (comment)](https://github.com/python-control/Slycot/pull/206#issuecomment-1879847670)

Splitting it out is okay as long as you don't duplicate code along with it. I'd rather keep it together as much as possible without losing the overview. So `test_aXNN.py` or even `test_analysis_aX.py`  might indeed be a good compromise.

**Comment by KybernetikJo** [comment]
I also could just drop all changes to pytest files.  We could do a reorganization in other pull requests, if necessary.

I think I have not added a single pytest to the project. I believe we would not lose anything.
It was only about reorganization, which was triggered by some misunderstanding of the docstring-pytest (info).
Actually, I was not aware about the whole test_analysis.py thing, when I started.

This would make the PR cleaner. It would be mainly about numpydoc changes, as the PR name suggests.

**Comment by bnavigator** [comment]
> This would make the PR cleaner. It would be mainly about numpydoc changes, as the PR name suggests.

That's a very good point. Please feel free to discuss/propose structure changes in another PR.

**Comment by bnavigator** [comment]
LGTM. Ready to merge?

**Comment by bnavigator** [comment]
Thank you very much for the hard work and the patience.

**Pull request merged**